### PR TITLE
ScanFilter Panel for Filtered Nodes

### PR DESCRIPTION
**Pull request by preetkaran20**

### CODE REVIEW

**Comment by preetkaran20** [comment]
New Screen 
image

**Comment by thc202** [comment]
This is also to include in 2.9.0, right?

**Comment by preetkaran20** [comment]
> This is also to include in 2.9.0, right?

Hi @thc202,
Yes i am working on this too. will resolve the conflicts today or tomorrow.

**Comment by preetkaran20** [comment]
Hi @thc202 and @kingthorin ,

I have done the changes. please review. 

thanks,
Karan

File: zap/src/main/java/org/parosproxy/paros/core/scanner/HostProcess.java [hunk 1]
```diff
@@ -507,6 +507,16 @@ private boolean filterNode(StructuralNode node) {
             try {
                 FilterResult filterResult = scanFilter.isFiltered(node);
                 if (filterResult.isFiltered()) {
+                    try {
+                        HttpMessage msg = node.getHistoryReference().getHttpMessage();
+                        parentScanner.notifyFilteredMessage(msg, filterResult.getReason());
+                    } catch (HttpMalformedHeaderException | DatabaseException e) {
+                        log.error(
```
Comment (thc202) [inline zap/src/main/java/org/parosproxy/paros/core/scanner/HostProcess.java]: This can be warning (also, for consistency with log at line 554).

File: zap/src/main/java/org/parosproxy/paros/core/scanner/ScannerListener.java [hunk 2]
```diff
@@ -41,4 +41,7 @@
 
     // ZAP: Added notifyNewMessage
     void notifyNewMessage(HttpMessage msg);
+
+    // Added to notify reason for filtering message from scanning.
```
Comment (thc202) [inline zap/src/main/java/org/parosproxy/paros/core/scanner/ScannerListener.java]: Either add to after the license header or change to JavaDoc (inline comments are not useful, the ones above are for legacy reasons).

File: zap/src/main/java/org/zaproxy/zap/extension/ascan/ActiveScan.java [hunk 3]
```diff
@@ -222,7 +225,8 @@ public void hostNewScan(int id, String hostAndPort, HostProcess hostThread) {}
     @Override
     public void hostProgress(int id, String hostAndPort, String msg, int percentage) {
         // Calculate the percentage based on the average of all of the host processes
-        // This is an approximation as different host process make significantly different times
+        // This is an approximation as different host process make significantly
```
Comment (thc202) [inline zap/src/main/java/org/zaproxy/zap/extension/ascan/ActiveScan.java]: Can be reverted.

File: zap/src/main/java/org/zaproxy/zap/extension/ascan/ActiveScan.java [hunk 4]
```diff
@@ -234,6 +238,18 @@ public void hostProgress(int id, String hostAndPort, String msg, int percentage)
         }
     }
 
+    @Override
+    public void filteredMessage(HttpMessage msg, String reason) {
+        try {
+            filterMessageTableModel.addScanResult(SessionStructure.getNodeName(msg), reason);
```
Comment (thc202) [inline zap/src/main/java/org/zaproxy/zap/extension/ascan/ActiveScan.java]: It can use the URI from the message instead of node name.

File: zap/src/main/java/org/zaproxy/zap/extension/ascan/ActiveScan.java [hunk 5]
```diff
@@ -234,6 +238,18 @@ public void hostProgress(int id, String hostAndPort, String msg, int percentage)
...
+        } catch (URIException e) {
+            log.error(
```
Comment (thc202) [inline zap/src/main/java/org/zaproxy/zap/extension/ascan/ActiveScan.java]: `warn` or `debug` (this shouldn't really happen, given that the message was already created processed).

File: zap/src/main/java/org/zaproxy/zap/extension/ascan/ActiveScanPanel.java [hunk 6]
```diff
@@ -66,17 +72,23 @@
     private static final String ZERO_NEW_ALERTS_LABEL_TEXT = "0";
 
     private static final ActiveScanTableModel EMPTY_RESULTS_MODEL = new ActiveScanTableModel();
+    private static final FilterMessageTableModel EMPTY_FILTER_MESSAGE_MODEL =
+            new FilterMessageTableModel();
 
     private ExtensionActiveScan extension;
     private JScrollPane jScrollPane;
+    private JScrollPane filterPane;
```
Comment (thc202) [inline zap/src/main/java/org/zaproxy/zap/extension/ascan/ActiveScanPanel.java]: This can be removed (and `jScrollPane`, now only used in the `getWorkPanel`).

File: zap/src/main/java/org/zaproxy/zap/extension/ascan/ActiveScanPanel.java [hunk 7]
```diff
@@ -89,6 +101,20 @@ public ActiveScanPanel(ExtensionActiveScan extension) {
                 "ascan",
                 new ImageIcon(ActiveScanPanel.class.getResource("/resource/icon/16/093.png")),
                 extension);
+        tabbedPane.addChangeListener(
+                new ChangeListener() {
+                    @Override
+                    public void stateChanged(ChangeEvent e) {
```
Comment (thc202) [inline zap/src/main/java/org/zaproxy/zap/extension/ascan/ActiveScanPanel.java]: Can use a lambda for shorter code.

File: zap/src/main/java/org/zaproxy/zap/extension/ascan/ActiveScanPanel.java [hunk 8]
```diff
@@ -229,19 +255,39 @@ public void clearFinishedScans() {
...
     @Override
-    protected JScrollPane getWorkPanel() {
-        if (jScrollPane == null) {
-            jScrollPane = new JScrollPane();
-            jScrollPane.setViewportView(getMessagesTable());
+    protected JPanel getWorkPanel() {
+        if (mainPanel == null) {
+            mainPanel = new JPanel(new BorderLayout());
+            tabbedPane = new JTabbedPane();
+            if (jScrollPane == null) {
```
Comment (thc202) [inline zap/src/main/java/org/zaproxy/zap/extension/ascan/ActiveScanPanel.java]: Should be removed.

File: zap/src/main/java/org/zaproxy/zap/extension/ascan/ActiveScanPanel.java [hunk 9]
```diff
@@ -229,19 +255,39 @@ public void clearFinishedScans() {
...
+                jScrollPane = new JScrollPane();
+                jScrollPane.setName("ActiveScanMessagePane");
+                jScrollPane.setViewportView(getMessagesTable());
+            }
+            if (filterPane == null) {
```
Comment (thc202) [inline zap/src/main/java/org/zaproxy/zap/extension/ascan/ActiveScanPanel.java]: Should be removed.

File: zap/src/main/java/org/zaproxy/zap/extension/ascan/ActiveScanPanel.java [hunk 10]
```diff
@@ -229,19 +255,39 @@ public void clearFinishedScans() {
...
     private void resetMessagesTable() {
         getMessagesTable().setModel(EMPTY_RESULTS_MODEL);
     }
 
-    private HistoryReferencesTable getMessagesTable() {
+    private void resetFilterMessageTable() {
+        getFilterMessageTable().setModel(EMPTY_FILTER_MESSAGE_MODEL);
+    }
+
+    private ZapTable getMessagesTable() {
```
Comment (thc202) [inline zap/src/main/java/org/zaproxy/zap/extension/ascan/ActiveScanPanel.java]: This can still return `HistoryReferencesTable`.

File: zap/src/main/java/org/zaproxy/zap/extension/ascan/FilterMessageTableModel.java [hunk 11]
```diff
@@ -0,0 +1,139 @@
...
+        }
+    }
+
+    /**
+     * Adds a new filtered message result.
+     *
+     * @param url
+     * @param reason for filtering message
+     */
+    public void addScanResult(String url, String reason) {
```
Comment (thc202) [inline zap/src/main/java/org/zaproxy/zap/extension/ascan/FilterMessageTableModel.java]: `addResult` or `addFilterResult`.

File: zap/src/main/java/org/zaproxy/zap/extension/ascan/FilterMessageTableModel.java [hunk 12]
```diff
@@ -0,0 +1,139 @@
...
+        FilteredMessageResult result = new FilteredMessageResult(url, reason);
+        filteredMessageResults.add(result);
+        fireTableRowsInserted(filteredMessageResults.size() - 1, filteredMessageResults.size() - 1);
+    }
+
+    /**
```
Comment (thc202) [inline zap/src/main/java/org/zaproxy/zap/extension/ascan/FilterMessageTableModel.java]: Could be removed (base class already has docs).

File: zap/src/main/java/org/zaproxy/zap/extension/ascan/FilterMessageTableModel.java [hunk 13]
```diff
@@ -0,0 +1,139 @@
...
+         * @param reason for filtering message
+         */
+        protected FilteredMessageResult(String url, String reason) {
+            super();
+            this.url = url;
+            this.reason = reason;
+        }
+
+        @Override
+        public int hashCode() {
```
Comment (thc202) [inline zap/src/main/java/org/zaproxy/zap/extension/ascan/FilterMessageTableModel.java]: Not necessary, same for equals.

File: zap/src/main/resources/org/zaproxy/zap/resources/Messages.properties [hunk 14]
```diff
@@ -563,6 +563,10 @@ ascan.toolbar.requests.label   = Num requests:
 ascan.toolbar.progress.label   = Progress:
 ascan.toolbar.progress.select  = --Select Scan--
 ascan.url.popup                = Active Scan single URL
+ascan.panel.tab.scannedMessages   = Scanned Messages
```
Comment (thc202) [inline zap/src/main/resources/org/zaproxy/zap/resources/Messages.properties]: `Messages Sent`?

File: zap/src/main/java/org/zaproxy/zap/extension/ascan/ActiveScanPanel.java [hunk 15]
```diff
@@ -250,6 +296,15 @@ private HistoryReferencesTable getMessagesTable() {
         return messagesTable;
     }
 
+    private ZapTable getFilterMessageTable() {
+        if (filterMessageTable == null) {
+            this.filterMessageTable = new ZapTable(EMPTY_FILTER_MESSAGE_MODEL);
```
Comment (thc202) [inline zap/src/main/java/org/zaproxy/zap/extension/ascan/ActiveScanPanel.java]: Ideally we should show the original message when the corresponding entry is selected, but something that can be added later (IMO).

File: zap/src/main/java/org/zaproxy/zap/extension/ascan/ActiveScan.java [hunk 16]
```diff
@@ -256,6 +272,10 @@ public void scannerComplete(int id) {
         return null;
     }
 
+    public FilterMessageTableModel getFilterMessageTableModel() {
```
Comment (thc202) [inline zap/src/main/java/org/zaproxy/zap/extension/ascan/ActiveScan.java]: Might worth leaving as package accessible for now.

File: zap/src/main/java/org/parosproxy/paros/core/scanner/ScannerListener.java (hunk 2 above)
Comment (preetkaran20) [reply zap/src/main/java/org/parosproxy/paros/core/scanner/ScannerListener.java]: sure.

File: zap/src/main/java/org/zaproxy/zap/extension/ascan/ActiveScan.java (hunk 3 above)
Comment (preetkaran20) [reply zap/src/main/java/org/zaproxy/zap/extension/ascan/ActiveScan.java]: done automatically by spotlessapply.

File: zap/src/main/java/org/parosproxy/paros/core/scanner/HostProcess.java (hunk 1 above)
Comment (preetkaran20) [reply zap/src/main/java/org/parosproxy/paros/core/scanner/HostProcess.java]: ok

File: zap/src/main/java/org/zaproxy/zap/extension/ascan/ActiveScan.java (hunk 5 above)
Comment (preetkaran20) [reply zap/src/main/java/org/zaproxy/zap/extension/ascan/ActiveScan.java]: ok

File: zap/src/main/java/org/zaproxy/zap/extension/ascan/ActiveScan.java (hunk 3 above)
Comment (thc202) [reply zap/src/main/java/org/zaproxy/zap/extension/ascan/ActiveScan.java]: Most likely because of a change done/reverted in the meantime, so this should be reverted as well.

File: zap/src/main/java/org/zaproxy/zap/extension/ascan/FilterMessageTableModel.java (hunk 13 above)
Comment (preetkaran20) [reply zap/src/main/java/org/zaproxy/zap/extension/ascan/FilterMessageTableModel.java]: ok

File: zap/src/main/resources/org/zaproxy/zap/resources/Messages.properties (hunk 14 above)
Comment (preetkaran20) [reply zap/src/main/resources/org/zaproxy/zap/resources/Messages.properties]: hmm ok

File: zap/src/main/java/org/zaproxy/zap/extension/ascan/ActiveScan.java (hunk 16 above)
Comment (preetkaran20) [reply zap/src/main/java/org/zaproxy/zap/extension/ascan/ActiveScan.java]: ok

File: zap/src/main/java/org/zaproxy/zap/extension/ascan/ActiveScan.java (hunk 3 above)
Comment (preetkaran20) [reply zap/src/main/java/org/zaproxy/zap/extension/ascan/ActiveScan.java]: oh ... but in case if modified in future, spotless will again do the same.

File: zap/src/main/resources/org/zaproxy/zap/resources/Messages.properties (hunk 14 above)
Comment (thc202) [reply zap/src/main/resources/org/zaproxy/zap/resources/Messages.properties]: I'm suggesting "sent" instead of "scanned" as the latter could be interpreted as the original scanned messages.

File: zap/src/main/java/org/zaproxy/zap/extension/ascan/ActiveScan.java (hunk 4 above)
Comment (preetkaran20) [reply zap/src/main/java/org/zaproxy/zap/extension/ascan/ActiveScan.java]: so can i use msg.getRequestHeader().getURI().toString() or shall i again call SessionStructure.getHostName(msg) ?

Comment (kingthorin) [reply zap/src/main/java/org/zaproxy/zap/extension/ascan/ActiveScan.java]: I believe he meant `msg.getRequestHeader().getURI().toString()`

File: zap/src/main/java/org/zaproxy/zap/extension/ascan/ActiveScanPanel.java (hunk 6 above)
Comment (preetkaran20) [reply zap/src/main/java/org/zaproxy/zap/extension/ascan/ActiveScanPanel.java]: you are saying this because of mainpanel which is not null so getWorkpanel will not even again initialise filterpane so no need to make it instance variable ?

File: zap/src/main/java/org/zaproxy/zap/extension/ascan/ActiveScan.java (hunk 4 above)
Comment (preetkaran20) [reply zap/src/main/java/org/zaproxy/zap/extension/ascan/ActiveScan.java]: sure thanks.

File: zap/src/main/java/org/zaproxy/zap/extension/ascan/ActiveScan.java (hunk 3 above)
Comment (thc202) [reply zap/src/main/java/org/zaproxy/zap/extension/ascan/ActiveScan.java]: You sure? I just reverted the change and `spotlessApply` didn't reformat it.

File: zap/src/main/java/org/zaproxy/zap/extension/ascan/ActiveScan.java (hunk 4 above)
Comment (thc202) [reply zap/src/main/java/org/zaproxy/zap/extension/ascan/ActiveScan.java]: Right.

File: zap/src/main/resources/org/zaproxy/zap/resources/Messages.properties (hunk 14 above)
Comment (preetkaran20) [reply zap/src/main/resources/org/zaproxy/zap/resources/Messages.properties]: ok so it will be names as `Sent Messages` ?

File: zap/src/main/java/org/zaproxy/zap/extension/ascan/ActiveScanPanel.java (hunk 10 above)
Comment (preetkaran20) [reply zap/src/main/java/org/zaproxy/zap/extension/ascan/ActiveScanPanel.java]: hmm correct.

File: zap/src/main/java/org/zaproxy/zap/extension/ascan/FilterMessageTableModel.java (hunk 13 above)
Comment (preetkaran20) [reply zap/src/main/java/org/zaproxy/zap/extension/ascan/FilterMessageTableModel.java]: ok

File: zap/src/main/java/org/zaproxy/zap/extension/ascan/FilterMessageTableModel.java (hunk 12 above)
Comment (preetkaran20) [reply zap/src/main/java/org/zaproxy/zap/extension/ascan/FilterMessageTableModel.java]: ok

File: zap/src/main/java/org/zaproxy/zap/extension/ascan/ActiveScan.java (hunk 3 above)
Comment (preetkaran20) [reply zap/src/main/java/org/zaproxy/zap/extension/ascan/ActiveScan.java]: oh .. but it does. let me try again

File: zap/src/main/resources/org/zaproxy/zap/resources/Messages.properties (hunk 14 above)
Comment (thc202) [reply zap/src/main/resources/org/zaproxy/zap/resources/Messages.properties]: Yeah.

File: zap/src/main/java/org/zaproxy/zap/extension/ascan/ActiveScanPanel.java (hunk 9 above)
Comment (preetkaran20) [reply zap/src/main/java/org/zaproxy/zap/extension/ascan/ActiveScanPanel.java]: sure

**Comment by preetkaran20** [comment]
Hi @thc202 and @kingthorin ,

done the changes. please approve.

File: zap/src/main/java/org/zaproxy/zap/extension/ascan/ActiveScanPanel.java (hunk 6 above)
Comment (thc202) [reply zap/src/main/java/org/zaproxy/zap/extension/ascan/ActiveScanPanel.java]: Yes.

**Comment by preetkaran20** [comment]
Hi @thc202 ,

it is also failing because of zap/build/reports/japicmp/japi.html

please help me fix this.

**Comment by thc202** [comment]
Apply the following to `zap.gradle.kts`:
```diff         "org.zaproxy.zap.extension.autoupdate.ExtensionAutoUpdate#getLatestVersionInfo(org.zaproxy.zap.extension.autoupdate.CheckForUpdateCallback)",
-        "org.zaproxy.zap.extension.autoupdate.ManageAddOnsDialog#checkForUpdates()"
+        "org.zaproxy.zap.extension.autoupdate.ManageAddOnsDialog#checkForUpdates()",
+        // Safe default method
+        "org.parosproxy.paros.core.scanner.ScannerListener#filteredMessage(org.parosproxy.paros.network.HttpMessage,java.lang.String)",
+        // Internal API
+        "org.zaproxy.zap.extension.ascan.ActiveScanPanel#getWorkPanel()"
     )
```

**Comment by preetkaran20** [comment]
> Apply the following to `zap.gradle.kts`:
> 
> ```diff
> -        "org.zaproxy.zap.extension.autoupdate.ManageAddOnsDialog#checkForUpdates()"
> +        "org.zaproxy.zap.extension.autoupdate.ManageAddOnsDialog#checkForUpdates()",
> +        // Safe default method
> +        "org.parosproxy.paros.core.scanner.ScannerListener#filteredMessage(org.parosproxy.paros.network.HttpMessage,java.lang.String)",
> +        // Internal API
> +        "org.zaproxy.zap.extension.ascan.ActiveScanPanel#getWorkPanel()"
>      )
> ```

thanks alot. done the changes. please review.

**Comment by thc202** [comment]
Thanks!

**Comment by kingthorin** [comment]
Ugh, missed squashing this at merge 🙁
